lib/stagger/tcp_client.js

```diff
@@ -0,0 +1,137 @@
+var events = require("events");
+var net = require("net");
+var util = require("util");
+
+var Parser = require('../protocol/parser');
+var encode = require('../protocol/encoder');
+
+// Manages a TCP connection to stagger, including reconnects
+// (De)serialises incoming/outgoing messages
+function TcpClient(tags, connectionOpts, timeout) {
+  events.EventEmitter.call(this);
+
+  this.timeout = timeout || 60000;
+  this.connected = false;
+
+  this.parser = new Parser(
+    function(method, body) {
+      switch (method) {
+        case 'pair:ping':
+          this.pong();
+          break;
+        case 'pair:pong':
+          break; // NOOP
+        default:
+          this.emit("message", method, body);
+          break;
+      }
+    }.bind(this),
+    function(err) {
+      console.log("Parse error: " + err);
+      this.socket.end(); // will trigger a close event, which will lead to a reconnect
+      this.handleShutdown();
+      this.parser.reset();
+    }.bind(this)
+  );
+
+  this.connect(tags, connectionOpts);
+}
+
+util.inherits(TcpClient, events.EventEmitter);
+var prototype = TcpClient.prototype;
+
+prototype.send = function(method, body) {
+  this.socket.write(encode(method, body));
+};
+
+// private
+
+prototype.connect = function(tags, connectionOpts) {
+  var socket = net.connect(connectionOpts, function() {
+    // on connected
+    if (!this.connected) {
+      this.connected = true;
+      this.emit("connected");
+      this.setupActivityCheck();
+    }
+    this.send("register_process", { "Tags": tags });
+  }.bind(this));
+
+  socket.on("data", function(buffer) {
+    this.parser.feedData(buffer);
+    this.resetActivity();
+  }.bind(this));
+
+  socket.on("end", function() {
+    console.log("Socket closed by peer");
+    // socket will have been destroyed because half open connections are not allowed
+  }.bind(this));
+
+  socket.on("timeout", function() {
+    console.log("Socket timeout");
+    socket.end(); // will trigger a close event
+  }.bind(this));
+
+  socket.on("error", function(error) {
+    console.log("Socket error: " + error);
+    socket.destroy();
+    // close event will be emitted
+  }.bind(this));
+
+  socket.on("close", function(isError) {
+    if (isError) {
+      console.log("Connection closed with error, reconnecting");
+    } else {
+      console.log("Connection closed, reconnecting");
+    }
+    this.handleShutdown();
+    this.reconnect(tags, connectionOpts);
+  }.bind(this));
+
+  this.socket = socket;
+};
+
+prototype.reconnect = function (tags, connectionOpts) {
+  setTimeout(
+    function() {
+      this.connect(tags, connectionOpts);
+    }.bind(this),
+    1000
+  );
+};
+
+prototype.ping = function() {
+  this.send("pair:ping", {});
+};
+
+prototype.pong = function() {
+  this.send("pair:pong", {});
+};
+
+prototype.resetActivity = function() {
+  this.activityAt = Date.now();
+};
+
+prototype.setupActivityCheck = function() {
```

Comment:
See an example (using emitted events, how appropriate for node!) https://github.com/pusher/em-hiredis/blob/master/lib/em-hiredis/support/inactivity_checker.rb